services/worker-manager/src/api.js

```diff
@@ -335,3 +338,79 @@ builder.declare({
   }
   return res.reply(result);
 });
+
+builder.declare({
+  method: 'get',
+  route: '/worker/register',
+  name: 'registerWorker',
+  title: 'Register a running worker',
+  stability: APIBuilder.stability.experimental,
+  input: 'register-worker-request.yml',
+  output: 'register-worker-response.yml',
+  description: [
+    'Register a running worker.  Workers call this method on worker start-up.',
+    '',
+    'This call both marks the worker as running and returns the credentials',
+    'the worker will require to perform its work.  The worker must provide',
+    'some proof of its identity, and that proof varies by provider type.',
+  ].join('\n'),
+}, async function(req, res) {
+  const {workerPoolId, providerId, workerGroup, workerId, workerIdentityProof} = req.body;
+
+  // carefully check each value provided, since we have not yet validated the
+  // worker's "proof"
+
+  const workerPool = await this.WorkerPool.load({workerPoolId}, true);
+  if (!workerPool) {
+    return res.reportError('ResourceNotFound',
+      `Worker pool ${workerPoolId} does not exist`, {});
+  }
+
+  const provider = this.providers.get(providerId);
+  if (!provider) {
+    return res.reportError('ResourceNotFound',
+      `Provider ${providerId} does not exist`, {});
+  }
+
+  if (workerPool.providerId !== providerId && !workerPool.previousProviderIds.includes(providerId)) {
+    return res.reportError('InputError',
+      `Worker pool ${workerPoolId} not associated with provider ${providerId}`, {});
+  }
+
+  const worker = await this.Worker.load({workerPoolId, workerGroup, workerId}, true);
+  if (!worker) {
+    return res.reportError('ResourceNotFound',
+      `Worker ${workerGroup}/${workerId} in worker pool ${workerPoolId} does not exist`, {});
+  }
+
+  if (worker.providerId !== providerId) {
+    return res.reportError('InputError',
+      `Worker ${workerGroup}/${workerId} does not have provider ${providerId}`, {});
+  }
+
+  const {expires, errorMessage} = await provider.registerWorker({worker, workerPool, workerIdentityProof});
+
+  if (errorMessage) {
```

Comment:
I actually liked the behavior before of throwing errors, catching, logging them, and then returning a generic message. 

This way seems easier to get wrong and has less useful error messages in our logs.

Reply:
It's not a great design, you're right.  I'm looking for a way to distinguish 500's and 400's, basically, with some useful generic messages for the 400's.  If I switched it to use a dedicated error class, would that be good?

Reply:
I wonder if we should intentionally _not_ distinguish 500/400 here. Otherwise a dedicated class definitely makes sense to me!

Reply:
I did that :)

Reply:
Sorry, that comment didn't appear.  One distinction is that 500 is implicitly a server-side issue and will likely be retried, while a 400 is a pretty clear "ya done something wrong" indication.